### PR TITLE
Use common code for tensor product polynomial evaluation

### PR DESCRIPTION
**Pull request by kronbichler**
Looking at #15915 and the Raviart-Thomas elements, I observed that we have a rather clumsy implementation of `AnisotropicPolynomials::evaluate` because it loads the same element over and over again, because the compiler cannot understand the aliasing. Rather than creating temporary variables, I decided that it is better to unroll the tensor product indices of the complete space in a dimension-independent way and use the same code for `TensorProductPolynomials` and `AnisotropicPolynomials`.

### CODE REVIEW

**Comment by kronbichler** [comment]
It turns out that my modifications did not uniformly make the situation better, because my new code adds some indirect addressing. This is addressed by tabulating all dimensions except the innermost one, which we go through in a separate loop. This makes the code more lengthy, but we now benefit from the fact that this only needs to be done once.

File: tests/base/polynomial_test.cc [hunk 1]
```diff
@@ -34,7 +34,7 @@ void
 check_poly(const Point<dim> &x, const PolynomialType &p)
 {
   const unsigned int          n   = p.n();
-  const double                eps = 5.0e-15;
+  const double                eps = 1.0e-14;
```
Comment (blaisb) [inline tests/base/polynomial_test.cc:37]: Curiosity, why the change here?

**Review by blaisb — APPROVED** [review]
Just a quick question but that's all. Very cool code, actually very readable for that kind of implementation.
Thanks!

File: tests/base/polynomial_test.cc (hunk 1 above)
Comment (blaisb) [reply tests/base/polynomial_test.cc:37]: Awesome. Thanks for clarifying it up :). Very informative.

File: source/base/tensor_product_polynomials.cc [hunk 2]
```diff
@@ -303,6 +303,192 @@ TensorProductPolynomials<0, Polynomials::Polynomial<double>>::compute_grad_grad(
...
+      std::vector<Tensor<4, dim>> &    fourth_derivatives)
+    {
+      const bool update_values = (values.size() == indices.size() * size_x),
+                 update_grads  = (grads.size() == indices.size() * size_x),
+                 update_grad_grads =
+                   (grad_grads.size() == indices.size() * size_x),
+                 update_3rd_derivatives =
+                   (third_derivatives.size() == indices.size() * size_x),
+                 update_4th_derivatives =
+                   (fourth_derivatives.size() == indices.size() * size_x);
```
Comment (masterleinad) [inline source/base/tensor_product_polynomials.cc:339]: I think we avoid using this style.
```suggestion
      const bool update_values = (values.size() == indices.size() * size_x);
      const bool update_grads  = (grads.size() == indices.size() * size_x);
      const bool update_grad_grads =
                   (grad_grads.size() == indices.size() * size_x);
      const bool update_3rd_derivatives =
                   (third_derivatives.size() == indices.size() * size_x);
      const bool update_4th_derivatives =
                   (fourth_derivatives.size() == indices.size() * size_x);
```

**Review by masterleinad — APPROVED** [review]
